### PR TITLE
[doc] feat: Add fine-grained profiling tutorial for FSDP and Megatron on Ascend

### PR DESCRIPTION
**Pull request by mengchengTang**
### What does this PR do?

This PR introduces a comprehensive guide for fine-grained performance profiling on Ascend devices, supporting both FSDP and Megatron backends. To address the challenge of massive data volume during full profiling in large-scale training, this tutorial implements a "Key Path Sampling" strategy.
Key Highlights:
**Rollout Stage:** Detailed instructions for profiling vLLM and SGLang inference engines using torch_npu.profiler schedules.
**Training Stage:** Code instrumentation examples for compute_log_prob and update_policy phases to capture specific micro-batches or mini-batches.
**Backend Specifics:** Differentiated guidance for FSDP (Micro-Batch level control) and Megatron (Mini-Batch level control).

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

**Comment by tardis-key** [comment]
maintain the English version synchronously?

File: docs/ascend_tutorial/ascend_profiling_zh.rst [hunk 1]
```diff
@@ -76,14 +77,17 @@ Last updated: 08/14/2025.
...
       actor_rollout_ref:
-         actor:
+         actor:  # 设置 actor role 的 profiler 采集配置参数
             profiler:
                enable: True
                all_ranks: True
                tool_config:
                   npu:
                      discrete: False
+                     contents: [npu, cpu]
```
Comment (FightingZhen) [inline docs/ascend_tutorial/ascend_profiling_zh.rst]: 这里建议通过注释，说明配置这个参数的实际含义是什么

File: docs/ascend_tutorial/ascend_profiling_zh.rst [hunk 2]
```diff
@@ -12,7 +13,7 @@ Last updated: 08/14/2025.
 
 使用两级profile设置来控制数据采集
 
-- 全局采集控制：使用verl/trainer/config/ppo_trainer.yaml中的配置项控制采集的模式和步数，
+- 全局采集控制：使用verl/trainer/config/ppo_trainer.yaml(FSDP)或verl/trainer/config/ppo_megatron_trainer.yaml(MindSpeed)中的配置项控制采集的模式和步数，
```
Comment (FightingZhen) [inline docs/ascend_tutorial/ascend_profiling_zh.rst]: 建议“或”前面加一个逗号，分割一下语句

File: docs/ascend_tutorial/ascend_profiling_zh.rst [hunk 3]
```diff
@@ -123,4 +128,241 @@ Last updated: 08/14/2025.
...
+--------------------
+
+背景与挑战
+~~~~~~~~~~
+
+上述基于配置文件的采集方式虽然便捷，但在 **长序列 (Long Context)** 或 **大全局批量 (Large Global Batch Size)** 的训练场景中面临挑战。
+在一个完整的训练步 (Step) 内，模型计算呈现出高频次、重复性的特征：
+
+1. Rollout 阶段：序列生成 (Generate Sequence) 是一个自回归过程，涉及成千上万次 Decoder 模型的前向计算。
+2. Training 阶段：为了控制显存峰值，VeRL 通常采用 Micro-Batch 策略，将庞大的数据流切分为多个微批次进行计算。
```
Comment (FightingZhen) [inline docs/ascend_tutorial/ascend_profiling_zh.rst]: 统一采用全小写的"verl"

File: docs/ascend_tutorial/ascend_profiling_zh.rst [hunk 4]
```diff
@@ -123,4 +128,241 @@ Last updated: 08/14/2025.
...
+   
+  - compute_log_prob (Actor/Ref)：涉及多轮纯前向传播。
+  - update_policy (Actor/Critic)：涉及多轮前向与反向传播。
+
+这种特性会导致全量 Profiling 产生海量且重复的算子记录。如下图所示：
+
+.. image:: https://raw.githubusercontent.com/mengchengTang/verl-data/master/verl_ascend_profiler.png
```
Comment (FightingZhen) [inline docs/ascend_tutorial/ascend_profiling_zh.rst:152]: 关于图片的建议：
1、名称是否考虑写全，比如compute_log_prob？
2、是否考虑规范化三方库的名称大小写，比如vLLM和SGLang

File: docs/ascend_tutorial/ascend_profiling_zh.rst [hunk 5]
```diff
@@ -123,4 +128,241 @@ Last updated: 08/14/2025.
...
+      class DataParallelPPOActor(BasePPOActor):
+  
+          def compute_log_prob(self, data: DataProto, calculate_entropy=False) -> torch.Tensor:
+          
+  +           # 准备 profiler (配置同上，略)
+  +           self.prof_npu = torch_npu.profiler.profile(
+  +               # ...
+  +               # wait=0, warmup=0, active=1: 直接采集第一个 micro-batch
+  +               schedule=torch_npu.profiler.schedule(wait=0, warmup=0, active=1, repeat=1),
+  +               on_trace_ready=torch_npu.profiler.tensorboard_trace_handler("./outputs/actor_compute_log_prob", analyse_flag=True)
```
Comment (FightingZhen) [inline docs/ascend_tutorial/ascend_profiling_zh.rst]: 下面的逻辑，默认是按ref角色采集，这里默认是不是也设置为ref角色的文件夹名字？最好是按角色自动区分文件夹名称

File: docs/ascend_tutorial/ascend_profiling_zh.rst [hunk 6]
```diff
@@ -123,4 +128,241 @@ Last updated: 08/14/2025.
...
+  +           if is_ref:
+  +               self.prof_npu.start()
+  
+              for micro_batch in micro_batches:
+  
+                  # ... 原始计算逻辑 ...
+                  with torch.no_grad():
+                      entropy, log_probs = self._forward_micro_batch(...)
+                      
+  +                   # 驱动 schedule
```
Comment (FightingZhen) [inline docs/ascend_tutorial/ascend_profiling_zh.rst]: 上面的.step()没有加注释，看要不要统一加一下

File: docs/ascend_tutorial/ascend_profiling_zh.rst [hunk 7]
```diff
@@ -123,4 +128,241 @@ Last updated: 08/14/2025.
...
+      class MegatronPPOActor(BasePPOActor):
+          
+          def update_policy(self, dataloader: Iterable[DataProto]) -> dict:
+              # ...
+  +           # 准备 profiler (配置同上，略)
+  +           self.prof_npu = torch_npu.profiler.profile(
+  +               # ...
+  +               # 仅采集第一个 Mini Batch 的计算（含所有 Micro-Batch）和一次优化器更新
+  +               schedule=torch_npu.profiler.schedule(wait=0, warmup=0, active=1, repeat=1),
+  +               on_trace_ready=torch_npu.profiler.tensorboard_trace_handler("./outputs/megatron_update_profile", analyse_flag=True)
```
Comment (FightingZhen) [inline docs/ascend_tutorial/ascend_profiling_zh.rst]: 导出文件名和FSDP后端不一致，建议统一，比如FSDP的导出文件名叫fsdp_actor_update_profile，megatron的导出文件名叫megatron_actor_update_profile？

File: docs/ascend_tutorial/ascend_profiling_zh.rst (hunk 2 above)
Comment (mengchengTang) [reply docs/ascend_tutorial/ascend_profiling_zh.rst]: 已修改

File: docs/ascend_tutorial/ascend_profiling_zh.rst (hunk 1 above)
Comment (mengchengTang) [reply docs/ascend_tutorial/ascend_profiling_zh.rst]: 已添加实际含义，并提醒用户已有的配置项

File: docs/ascend_tutorial/ascend_profiling_zh.rst (hunk 3 above)
Comment (mengchengTang) [reply docs/ascend_tutorial/ascend_profiling_zh.rst]: 已修改为小写

File: docs/ascend_tutorial/ascend_profiling_zh.rst (hunk 4 above)
Comment (mengchengTang) [reply docs/ascend_tutorial/ascend_profiling_zh.rst:152]: 已按照建议修改

File: docs/ascend_tutorial/ascend_profiling_zh.rst (hunk 5 above)
Comment (mengchengTang) [reply docs/ascend_tutorial/ascend_profiling_zh.rst]: 已修改逻辑，文件名会区分ref和actor

File: docs/ascend_tutorial/ascend_profiling_zh.rst (hunk 6 above)
Comment (mengchengTang) [reply docs/ascend_tutorial/ascend_profiling_zh.rst]: 已统一补充

File: docs/ascend_tutorial/ascend_profiling_zh.rst (hunk 7 above)
Comment (mengchengTang) [reply docs/ascend_tutorial/ascend_profiling_zh.rst]: 已按照建议统一修改

File: docs/ascend_tutorial/ascend_profiling_zh.rst [hunk 8]
```diff
@@ -116,4 +116,239 @@ Last updated: 08/14/2025.
...
+
+                # ... 原始计算逻辑 ...
+                with torch.no_grad():
+                    entropy, log_probs = self._forward_micro_batch(...)
+                    
+                    # 驱动 schedule
+                    if is_ref:
+                        prof.step()
+                
+                # ...
```
[the remark this answers is not in the document]
Comment (mengchengTang) [reply docs/ascend_tutorial/ascend_profiling_zh.rst]: schedule配合prof.step会自动调用到prof.stop，此处不需要显示调用stop

File: docs/ascend_tutorial/ascend_profiling_zh.rst [hunk 9]
```diff
@@ -116,4 +116,239 @@ Last updated: 08/14/2025.
...
+
+                    for i, micro_batch in enumerate(micro_batches):
+                        # ... 原始 Forward & Backward 逻辑 ...
+                        # ... loss.backward() ...
+                        pass
+    
+                    grad_norm = self._optimizer_step()
+                    
+                    # 驱动 schedule
+                    prof.step()
```
[the remark this answers is not in the document]
Comment (mengchengTang) [reply docs/ascend_tutorial/ascend_profiling_zh.rst]: schedule配合prof.step会自动调用到prof.stop，此处不需要显示调用stop

File: docs/ascend_tutorial/ascend_profiling_zh.rst [hunk 10]
```diff
@@ -116,4 +116,239 @@ Last updated: 08/14/2025.
...
+            prof.start()
+            
+            for data in dataloader:
+                # ... 内部会调用 self.forward_backward_batch 进行计算 ...
+                # ... metric_micro_batch = self.forward_backward_batch(...)
+                
+                # ... self.actor_optimizer.step() ...
+                
+                # 驱动 schedule
+                prof.step()
```
[the remark this answers is not in the document]
Comment (mengchengTang) [reply docs/ascend_tutorial/ascend_profiling_zh.rst]: schedule配合prof.step会自动调用到prof.stop，此处不需要显示调用stop

**Comment by mengchengTang** [comment]
> maintain the English version synchronously?
